rfd/0014-session-2FA.md

```diff
@@ -0,0 +1,426 @@
+
+### 2nd factor
+
+There are a variety of MFA options available, but for this design we'll focus
+on U2F hardware tokens, because:
+
+- portability: U2F devices are supported on all major OSs and browsers (vs
+  TouchID or Windows Hello)
+- UX: tapping a USB token multiple times per day is low friction (vs typing in
+  TOTP codes)
+- availability: many engineers already own a U2F token (like a YubiKey), since
+  they are usable on popular websites (vs HSMs or smartcards)
+
+We may consider adding support for other MFA options, if there's demand.
+
+### U2F device management
+
+A prerequisite for usable 2FA integration is solid 2FA device management. This
+work is tracked separately, as [RFD 15](0015-2fa-management.md), to keep
+designs reasonably scoped and understandable.
+
+For this RFD, we assume that:
+- teleport 2FA device management is separate from SSO 2FA
+- teleport supports 2FA device management on CLI and web
+- a user can have multiple 2FA devices registered, including multiple security
+  tokens
+- a user can remove registered 2FA devices
+
+### authn protocol
+
+The design leverages short-lived SSH and TLS certificates per session. Cert
+expiry is used to limit the cert to a single "session".
+
+For all protocols, the flow is roughly:
+
+1. client requests a new certificate for the session
+2. client and server perform the U2F challenge exchange, with user tapping the
+   security token
+3. server issues a short-lived certificate with encoded constraints
+4. client uses the in-memory certificate to start the session and discards the
+   certificate
+
+The short-lived certificate is used for regular SSH or mTLS handshakes, with
+server validating it using the presented constraints.
+
+#### constraints
+
+Each session has the following constraints, encoded in the TLS or SSH
+certificate issued after MFA and enforced server-side:
+
+- cert expiry: each certificate is valid for 1min, during which the client can
+  _establish_ a session
+- session TTL: each session is terminated server-side after 30min, whether
+  active or idle
+  - this is important to prevent a compromised session from being artificially
+    kept alive forever, with some simulated activity
+- target: a specific server, k8s cluster, database or web app this session is for
+- client IP: only connections from the same IP that passed a MFA check can
+  establish a session
+
+### session UX
+
+UX is the same for all protocols: initiate session -> tap security key -> proceed.
+But the plumbing details are different:
+
+#### ssh
+
+The U2F handshake is performed by `tsh ssh`, before the actual SSH connection:
+
+```sh
+awly@localhost $ tsh ssh server1
+please tap your security key... <tap>
+awly@server1 #
+```
+
+For OpenSSH, `tsh ssh` can be injected using `ProxyCommand` option in the
+config, with identical UX.
+
+For the Web UI, the U2F exchange happens over the existing websocket
+connection, using JS messages (exact format TBD), before terminal traffic is
+allowed.
```

Comment:
We've a sub-set of customers who provide locked down Chrome books and only interact with Teleport using the UI. We should make sure this UX is a smooth as possible. e.g. inline message vs blocking modal

Reply:
We can print an inline message, but the interaction has to go via browser's UI.
The JS API for U2F doesn't let you suppress the browser UI component AFAIK.

Is that a problem with chromebooks?

Reply:
Thanks for the clarification. I don't think it'll be an issue but this reminds me that I should acquire a Chrome book so I can try the UX.